src/runtime/ndarray.cc

```diff
@@ -69,8 +69,14 @@ struct NDArray::Internal {
       // if the array is still pinned before freeing, unpin it.
       if (ptr->pinned_by_dgl_)
         UnpinContainer(ptr);
-      dgl::runtime::DeviceAPI::Get(ptr->dl_tensor.ctx)->FreeDataSpace(
-          ptr->dl_tensor.ctx, ptr->dl_tensor.data);
+      TensorDispatcher* td = TensorDispatcher::Global();
+      if (td->IsAvailable()) {
+        DeviceAPI::Get(ptr->dl_tensor.ctx)->FreeWorkspace(
+            ptr->dl_tensor.ctx, ptr->dl_tensor.data);
+      } else {
+        DeviceAPI::Get(ptr->dl_tensor.ctx)->FreeDataSpace(
+            ptr->dl_tensor.ctx, ptr->dl_tensor.data);
+      }
```

Comment:
I don't quite understand what's going on here.  What's the difference between FreeWorkspace and FreeDataSpace, and why we use FreeWorkspace if tensoradapter is available and FreeDataSpace otherwise?